lsat/satisfier.go

```diff
@@ -115,3 +119,64 @@ func NewCapabilitiesSatisfier(service string, targetCapability string) Satisfier
 		},
 	}
 }
+
+// NewTimeoutSatisfier checks if an LSAT is expired or not. The Satisfier takes
+// a service name to set as the condition prefix and currentTimestamp to
+// compare against the expiration(s) in the caveats. The expiration time is
+// retrieved from the caveat values themselves. The satisfier will also make
+// sure that each subsequent caveat of the same condition only has increasingly
+// strict expirations.
+func NewTimeoutSatisfier(service string, now func() time.Time) Satisfier {
+	return Satisfier{
+		Condition: service + CondTimeoutSuffix,
+		SatisfyPrevious: func(prev, cur Caveat) error {
+			prevValue, err := strconv.ParseInt(prev.Value, 10, 64)
+			if err != nil {
+				return fmt.Errorf("error parsing previous "+
+					"caveat value: %w", err)
+			}
+
+			currValue, err := strconv.ParseInt(cur.Value, 10, 64)
+			if err != nil {
+				return fmt.Errorf("error parsing caveat "+
+					"value: %w", err)
+			}
+
+			prevTime := time.Unix(prevValue, 0)
+			currTime := time.Unix(currValue, 0)
+
+			// Satisfier should fail if a previous timestamp in the
+			// list is earlier than ones after it b/c that means
+			// they are getting more permissive.
+			if prevTime.Before(currTime) {
+				return fmt.Errorf("%s caveat violates "+
+					"increasing restrictiveness",
+					service+CondTimeoutSuffix)
+			}
+
+			return nil
+		},
+		SatisfyFinal: func(c Caveat) error {
+			expirationTimestamp, err := strconv.ParseInt(
+				c.Value, 10, 64,
+			)
+			if err != nil {
+				return fmt.Errorf("caveat value not a valid "+
+					"integer: %v", err)
+			}
+
+			expirationTime := time.Unix(expirationTimestamp, 0)
+
+			// Make sure that the final relevant caveat is not
+			// passed the current date/time
+			if now().Before(expirationTime) {
+				return nil
+			}
+
+			return fmt.Errorf(
+				"not authorized to access service. LSAT has " +
```

Comment:
nit: since this is a returned error, this should be formatted like this:

```
	return fmt.Errorf("not authorized to access " +
				"service. LSAT has expired")
```